File: callings_diagram/src/generate.zig
const std = @import("std");
const Config = @import("config.zig").Config;
const data = @import("data.zig");

const DiagramGenerationError = error{
    OrgBubbleWidthTooSmall,
    UnrecognizedOrgName,
} || std.mem.Allocator.Error || std.fmt.AllocPrintError || data.DateFormatError;

pub fn diagram_file_contents(allocator: std.mem.Allocator, orgs: std.StringArrayHashMap(data.Organization), conf: Config) ![]const u8 {
    var file_contents = std.ArrayList(u8).init(allocator);

    var rng = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.microTimestamp())));
    const diagram_header = try std.fmt.allocPrint(allocator,
        \\<mxfile host="app.diagrams.net" agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.2 Safari/605.1.15" version="26.0.5">
        \\  <diagram id="{}" name="Page-1">
        \\    <mxGraphModel grid="1" page="0" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" pageScale="1" pageWidth="827" pageHeight="1169" math="0" shadow="0">
        \\      <root>
        \\        <mxCell id="0" />
        \\        <mxCell id="1" parent="0" />
    , .{rng.random().int(u128)});

    try file_contents.appendSlice(diagram_header);

    // TODO
    const min_org_bubble_height: i32 = conf.org_bubble_title_height + conf.calling_bubble_dimensions.height + 2 * conf.calling_bubble_vertical_margins;

    const calling_bubble_row_width_no_margins: i32 = conf.calling_bubble_dimensions.width * conf.calling_bubbles_per_row;
    const calling_bubble_left_margin_in_org_bubble: i32 = @divTrunc((conf.org_bubble_width - calling_bubble_row_width_no_margins), (conf.calling_bubbles_per_row + 1));

    const sub_org_bubble_width: i32 = conf.org_bubble_width - 2 * conf.sub_org_bubble_horzontal_margins;
    const calling_bubble_left_margin_in_sub_org_bubble: i32 = @divTrunc((sub_org_bubble_width - calling_bubble_row_width_no_margins), (conf.calling_bubbles_per_row + 1));

    if (calling_bubble_left_margin_in_sub_org_bubble < conf.calling_bubble_min_horizontal_margin) {
        return DiagramGenerationError.OrgBubbleWidthTooSmall;
    }

    var org_bubble_cursor_x: i32 = conf.diagram_start_x;
    var org_bubble_cursor_y: i32 = conf.diagram_start_y;

    var calling_num: i32 = 0;

    for (conf.org_ordering) |ordering| {
        const org_name = try sanitize(allocator, ordering.name);
        const start_new_column = ordering.begins_new_column;

        const org = orgs.getPtr(org_name) orelse return DiagramGenerationError.UnrecognizedOrgName;

        var org_bubble_id = try allocator.alloc(u8, org_name.len);
        _ = std.mem.replace(u8, org_name, " ", "-", org_bubble_id);
        org_bubble_id = try std.fmt.allocPrint(allocator, "{s}-{}", .{ org_bubble_id, rand_tag(&rng) });

        if (start_new_column) {
            org_bubble_cursor_x += conf.org_bubble_width + conf.org_bubble_margins.right + conf.org_bubble_margins.left;
            org_bubble_cursor_y = conf.diagram_start_y;
        }

        org_bubble_cursor_y += conf.org_bubble_margins.top;
        const org_bubble_x: i32 = org_bubble_cursor_x;
        const org_bubble_y: i32 = org_bubble_cursor_y;

        var org_calling_bubble_elems = std.ArrayList([]const u8).init(allocator);

        var calling_bubble_cursor_x: i32 = 0;
        var calling_bubble_cursor_y: i32 = 0;

        calling_bubble_cursor_y += conf.org_bubble_title_height + conf.calling_bubble_vertical_margins;

        var i: i32 = 0;
        for (org.callings.items) |calling| {
            const member = calling.member orelse continue;

            if (i != 0) {
                if (@mod(i, conf.calling_bubbles_per_row) == 0) {
                    calling_bubble_cursor_x = 0;

                    const calling_bubble_height_with_margin = conf.calling_bubble_dimensions.height + conf.calling_bubble_vertical_margins;
                    calling_bubble_cursor_y += calling_bubble_height_with_margin;
                } else {
                    calling_bubble_cursor_x += conf.calling_bubble_dimensions.width;
                }
            }

            calling_bubble_cursor_x += calling_bubble_left_margin_in_org_bubble;

            const calling_bubble_elem = try std.fmt.allocPrint(allocator,
                \\          <mxCell id="calling-{}" value="&lt;div&gt;&lt;b&gt;&lt;font style=&quot;font-size: 18px;&quot;&gt;{s}&lt;/font&gt;&lt;/b&gt;&lt;/div&gt;&lt;div&gt;&lt;br&gt;&lt;/div&gt;&lt;div&gt;{s}, {s}&lt;/div&gt;&lt;div&gt;Since: {s}&lt;/div&gt;" style="rounded=1;whiteSpace=wrap;html=1;align=left;spacingLeft=0;spacingTop=0;spacing=10;fontSize=16;" vertex="1" parent="{s}">
                \\            <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
                \\          </mxCell>
            , .{
                calling_num,
                try sanitize(allocator, calling.name),
                member.last_name,
                member.given_names,
                try member.held_calling_since.allocPrint(allocator),
                org_bubble_id,
                calling_bubble_cursor_x,
                calling_bubble_cursor_y,
                conf.calling_bubble_dimensions.width,
                conf.calling_bubble_dimensions.height,
            });
            calling_num += 1;

            try org_calling_bubble_elems.append(calling_bubble_elem);

            i += 1;
        }

        org_bubble_cursor_x = org_bubble_x;

        // Assumption: If an org has sub-orgs, there are no callings in the org that aren't part of
        //             a sub-org
        var sub_org_bubble_cursor_y: i32 = conf.org_bubble_title_height;
        var all_sub_orgs_in_org_bubble_elems = std.ArrayList(std.ArrayList([]const u8)).init(allocator);

        for (org.children.items) |child| {
            // If there are grandchildren, put their callings the sub org
            var callings = std.ArrayList(data.Calling).init(allocator);

            for (child.callings.items) |c| {
                if (c.member == null) {
                    continue;
                }
                try callings.append(c);
            }

            for (child.children.items) |grandchild| {
                for (grandchild.callings.items) |c| {
                    if (c.member == null) {
                        continue;
                    }
                    try callings.append(c);
                }
            }

            if (callings.items.len == 0) {
                continue;
            }

            const child_name = try sanitize(allocator, child.name);

            var sub_org_bubble_id = try allocator.alloc(u8, child_name.len);
            _ = std.mem.replace(u8, child_name, " ", "-", sub_org_bubble_id);
            sub_org_bubble_id = try std.fmt.allocPrint(allocator, "{s}-{}", .{ sub_org_bubble_id, rand_tag(&rng) });

            var sub_org_bubble_elems = std.ArrayList([]const u8).init(allocator);
            _ = try sub_org_bubble_elems.addOne(); // Placeholder for sub-org bubble itself

            calling_bubble_cursor_x = 0;
            calling_bubble_cursor_y = conf.org_bubble_title_height + conf.calling_bubble_vertical_margins;

            i = 0;
            for (callings.items) |calling| {
                const member = calling.member orelse continue;

                if (i != 0) {
                    if (@mod(i, conf.calling_bubbles_per_row) == 0) {
                        calling_bubble_cursor_x = 0;

                        const calling_bubble_height_with_margin = conf.calling_bubble_dimensions.height + conf.calling_bubble_vertical_margins;
                        calling_bubble_cursor_y += calling_bubble_height_with_margin;
                    } else {
                        calling_bubble_cursor_x += conf.calling_bubble_dimensions.width;
                    }
                }

                calling_bubble_cursor_x += calling_bubble_left_margin_in_sub_org_bubble;

                const calling_bubble_elem = try std.fmt.allocPrint(allocator,
                    \\            <mxCell id="calling-{}" value="&lt;div&gt;&lt;b&gt;&lt;font style=&quot;font-size: 18px;&quot;&gt;{s}&lt;/font&gt;&lt;/b&gt;&lt;/div&gt;&lt;div&gt;&lt;br&gt;&lt;/div&gt;&lt;div&gt;{s}, {s}&lt;/div&gt;&lt;div&gt;Since: {s}&lt;/div&gt;" style="rounded=1;whiteSpace=wrap;html=1;align=left;spacingLeft=0;spacingTop=0;spacing=10;fontSize=16;" vertex="1" parent="{s}">
                    \\              <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
                    \\            </mxCell>
                , .{
                    calling_num,
                    try sanitize(allocator, calling.name),
                    member.last_name,
                    member.given_names,
                    try member.held_calling_since.allocPrint(allocator),
                    sub_org_bubble_id,
                    calling_bubble_cursor_x,
                    calling_bubble_cursor_y,
                    conf.calling_bubble_dimensions.width,
                    conf.calling_bubble_dimensions.height,
                });
                calling_num += 1;

                try sub_org_bubble_elems.append(calling_bubble_elem);

                i += 1;
            }

            // The first element in sub_org_calling_bubble_elems is a placeholder for the sub-org bubble itself,
            // so subtract one from the length of the sub_org_bubble_elems list to do this calculation
            const sub_org_filled_callings_count: i32 = if (sub_org_bubble_elems.items.len - 1 > std.math.maxInt(i32)) std.math.maxInt(i32) else @intCast(sub_org_bubble_elems.items.len - 1);
            var sub_org_bubble_height: i32 = (@divTrunc(sub_org_filled_callings_count - 1, conf.calling_bubbles_per_row) + 1) * (conf.calling_bubble_dimensions.height + conf.calling_bubble_vertical_margins) + conf.calling_bubble_vertical_margins + conf.org_bubble_title_height;
            if (sub_org_bubble_height < min_org_bubble_height) {
                sub_org_bubble_height = min_org_bubble_height;
            }

            sub_org_bubble_cursor_y += conf.sub_org_bubble_vertical_margins;

            const sub_org_bubble_elem = try std.fmt.allocPrint(allocator,
                \\          <mxCell id="{s}" value="&lt;font style=&quot;font-size: 22px;&quot;&gt;{s}&lt;/font&gt;" style="swimlane;whiteSpace=wrap;html=1;rounded=1;strokeWidth=4;startSize=40;strokeColor=#9E9E9E;fontColor=#6B6B6B;" vertex="1" parent="{s}">
                \\            <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
                \\          </mxCell>
            , .{
                sub_org_bubble_id,
                child_name,
                org_bubble_id,
                conf.sub_org_bubble_horzontal_margins,
                sub_org_bubble_cursor_y,
                sub_org_bubble_width,
                sub_org_bubble_height,
            });

            sub_org_bubble_elems.items[0] = sub_org_bubble_elem;

            sub_org_bubble_cursor_y += sub_org_bubble_height;

            try all_sub_orgs_in_org_bubble_elems.append(sub_org_bubble_elems);
        }

        const org_filled_callings_count: i32 = if (org_calling_bubble_elems.items.len > std.math.maxInt(i32)) std.math.maxInt(i32) else @intCast(org_calling_bubble_elems.items.len);
        var org_bubble_height: i32 = 0;

        if (org.children.items.len == 0) {
            org_bubble_height = (@divTrunc(org_filled_callings_count - 1, conf.calling_bubbles_per_row) + 1) * (conf.calling_bubble_dimensions.height + conf.calling_bubble_vertical_margins) + conf.calling_bubble_vertical_margins + conf.org_bubble_title_height;
        } else {
            org_bubble_height = sub_org_bubble_cursor_y + conf.sub_org_bubble_vertical_margins;
        }

        if (org_bubble_height < min_org_bubble_height) {
            org_bubble_height = min_org_bubble_height;
        }

        org_bubble_cursor_y += org_bubble_height;

        const org_bubble_elem = try std.fmt.allocPrint(allocator,
            \\        <mxCell id="{s}" value="&lt;font style=&quot;font-size: 22px;&quot;&gt;{s}&lt;/font&gt;" style="swimlane;whiteSpace=wrap;html=1;rounded=1;strokeWidth=4;startSize=40;" vertex="1" parent="1">
            \\          <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
            \\        </mxCell>
        , .{
            org_bubble_id,
            org_name,
            org_bubble_x,
            org_bubble_y,
            conf.org_bubble_width,
            org_bubble_height,
        });

        try file_contents.appendSlice(org_bubble_elem);

        for (org_calling_bubble_elems.items) |elem| {
            try file_contents.appendSlice(elem);
        }

        for (all_sub_orgs_in_org_bubble_elems.items) |sub_org_bubble_elems| {
            for (sub_org_bubble_elems.items) |elem| {
                try file_contents.appendSlice(elem);
            }
        }

        org_bubble_cursor_y += conf.org_bubble_margins.bottom;
    }

    const diagram_footer = try std.fmt.allocPrint(allocator,
        \\      </root>
        \\    </mxGraphModel>
        \\  </diagram>
        \\</mxfile>
    , .{});

    try file_contents.appendSlice(diagram_footer);

    return file_contents.items;
}

fn sanitize(allocator: std.mem.Allocator, str: []const u8) std.mem.Allocator.Error![]u8 {
    const sanitized = try allocator.alloc(u8, std.mem.replacementSize(u8, str, "&", "and"));
    _ = std.mem.replace(u8, str, "&", "and", sanitized);
    return sanitized;
}

fn rand_tag(rng: *std.rand.DefaultPrng) u32 {
    return rng.random().int(u32) % 900000 + 100000;
}

File: callings_diagram/src/generate.rs
use crate::config::Config;
use crate::data::Organization;
use rand::Rng;
use std::collections::HashMap;
use std::fmt;

#[derive(Debug)]
pub enum GenerateError {
    DataError(crate::data::DataError),
    InvalidLayout(String),
}

impl std::error::Error for GenerateError {}

impl fmt::Display for GenerateError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GenerateError::DataError(e) => write!(f, "Data error: {}", e),
            GenerateError::InvalidLayout(msg) => write!(f, "Invalid layout: {}", msg),
        }
    }
}

impl From<crate::data::DataError> for GenerateError {
    fn from(err: crate::data::DataError) -> Self {
        GenerateError::DataError(err)
    }
}

pub fn diagram_file_contents(
    orgs: &HashMap<String, Organization>,
    conf: &Config,
) -> Result<String, GenerateError> {
    let mut file_contents = String::new();

    let mut rng = rand::thread_rng();
    let diagram_id = rng.gen_range(0..u128::MAX);

    let diagram_header = format!(
        r#"<mxfile host="app.diagrams.net" agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.2 Safari/605.1.15" version="26.0.5">
  <diagram id="{}" name="Page-1">
    <mxGraphModel grid="1" page="0" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" pageScale="1" pageWidth="827" pageHeight="1169" math="0" shadow="0">
      <root>
        <mxCell id="0" />
        <mxCell id="1" parent="0" />"#,
        diagram_id
    );

    file_contents.push_str(&diagram_header);

    let min_org_bubble_height = conf.org_bubble_title_height
        + conf.calling_bubble_dimensions.height
        + 2 * conf.calling_bubble_vertical_margins;

    let calling_bubble_row_width_no_margins =
        conf.calling_bubble_dimensions.width * conf.calling_bubbles_per_row;
    let calling_bubble_left_margin_in_org_bubble = (conf.org_bubble_width
        - calling_bubble_row_width_no_margins)
        / (conf.calling_bubbles_per_row + 1);

    let sub_org_bubble_width = conf.org_bubble_width - 2 * conf.sub_org_bubble_horzontal_margins;
    let calling_bubble_left_margin_in_sub_org_bubble = (sub_org_bubble_width
        - calling_bubble_row_width_no_margins)
        / (conf.calling_bubbles_per_row + 1);

    if calling_bubble_left_margin_in_sub_org_bubble < conf.calling_bubble_min_horizontal_margin {
        return Err(GenerateError::InvalidLayout(
            "Org bubble width too small".to_string(),
        ));
    }

    let mut org_bubble_cursor_x = conf.diagram_start_x;
    let mut org_bubble_cursor_y = conf.diagram_start_y;

    let mut calling_num = 0;

    for ordering in &conf.org_ordering {
        let org_name = sanitize(&ordering.name);
        let start_new_column = ordering.begins_new_column;

        let org = orgs.get(&ordering.name).ok_or_else(|| {
            GenerateError::InvalidLayout(format!("Unrecognized org name: {}", ordering.name))
        })?;

        let mut org_bubble_id = org_name.replace(" ", "-");
        org_bubble_id = format!("{}-{}", org_bubble_id, rand_tag(&mut rng));

        if start_new_column {
            org_bubble_cursor_x += conf.org_bubble_width
                + conf.org_bubble_margins.right
                + conf.org_bubble_margins.left;
            org_bubble_cursor_y = conf.diagram_start_y;
        }

        org_bubble_cursor_y += conf.org_bubble_margins.top;
        let org_bubble_x = org_bubble_cursor_x;
        let org_bubble_y = org_bubble_cursor_y;

        let mut org_calling_bubble_elems = Vec::new();

        let mut calling_bubble_cursor_x = 0;
        let mut calling_bubble_cursor_y = 0;

        calling_bubble_cursor_y +=
            conf.org_bubble_title_height + conf.calling_bubble_vertical_margins;

        let mut i = 0;
        for calling in &org.callings {
            let member = match &calling.member {
                Some(m) => m,
                None => continue,
            };

            if i != 0 {
                if i % conf.calling_bubbles_per_row == 0 {
                    calling_bubble_cursor_x = 0;
                    let calling_bubble_height_with_margin = conf.calling_bubble_dimensions.height
                        + conf.calling_bubble_vertical_margins;
                    calling_bubble_cursor_y += calling_bubble_height_with_margin;
                } else {
                    calling_bubble_cursor_x += conf.calling_bubble_dimensions.width;
                }
            }

            calling_bubble_cursor_x += calling_bubble_left_margin_in_org_bubble;

            let calling_bubble_elem = format!(
                r#"          <mxCell id="calling-{}" value="&lt;div&gt;&lt;b&gt;&lt;font style=&quot;font-size: 18px;&quot;&gt;{}&lt;/font&gt;&lt;/b&gt;&lt;/div&gt;&lt;div&gt;&lt;br&gt;&lt;/div&gt;&lt;div&gt;{}, {}&lt;/div&gt;&lt;div&gt;Since: {}&lt;/div&gt;" style="rounded=1;whiteSpace=wrap;html=1;align=left;spacingLeft=0;spacingTop=0;spacing=10;fontSize=16;" vertex="1" parent="{}">
            <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
          </mxCell>"#,
                calling_num,
                sanitize(&calling.name),
                member.last_name,
                member.given_names,
                member.held_calling_since.to_string()?,
                org_bubble_id,
                calling_bubble_cursor_x,
                calling_bubble_cursor_y,
                conf.calling_bubble_dimensions.width,
                conf.calling_bubble_dimensions.height,
            );
            calling_num += 1;

            org_calling_bubble_elems.push(calling_bubble_elem);

            i += 1;
        }

        org_bubble_cursor_x = org_bubble_x;

        // Assumption: If an org has sub-orgs, there are no callings in the org that aren't part of
        //             a sub-org
        let mut sub_org_bubble_cursor_y = conf.org_bubble_title_height;
        let mut all_sub_orgs_in_org_bubble_elems = Vec::new();

        for child in &org.children {
            // If there are grandchildren, put their callings the sub org
            let mut callings = Vec::new();

            for c in &child.callings {
                if c.member.is_some() {
                    callings.push(c.clone());
                }
            }

            for grandchild in &child.children {
                for c in &grandchild.callings {
                    if c.member.is_some() {
                        callings.push(c.clone());
                    }
                }
            }

            if callings.is_empty() {
                continue;
            }

            let child_name = sanitize(&child.name);

            let mut sub_org_bubble_id = child_name.replace(" ", "-");
            sub_org_bubble_id = format!("{}-{}", sub_org_bubble_id, rand_tag(&mut rng));

            let mut sub_org_bubble_elems = Vec::new();
            sub_org_bubble_elems.push(String::new()); // Placeholder for sub-org bubble itself

            calling_bubble_cursor_x = 0;
            calling_bubble_cursor_y =
                conf.org_bubble_title_height + conf.calling_bubble_vertical_margins;

            i = 0;
            for calling in &callings {
                let member = match &calling.member {
                    Some(m) => m,
                    None => continue,
                };

                if i != 0 {
                    if i % conf.calling_bubbles_per_row == 0 {
                        calling_bubble_cursor_x = 0;
                        let calling_bubble_height_with_margin =
                            conf.calling_bubble_dimensions.height
                                + conf.calling_bubble_vertical_margins;
                        calling_bubble_cursor_y += calling_bubble_height_with_margin;
                    } else {
                        calling_bubble_cursor_x += conf.calling_bubble_dimensions.width;
                    }
                }

                calling_bubble_cursor_x += calling_bubble_left_margin_in_sub_org_bubble;

                let calling_bubble_elem = format!(
                    r#"            <mxCell id="calling-{}" value="&lt;div&gt;&lt;b&gt;&lt;font style=&quot;font-size: 18px;&quot;&gt;{}&lt;/font&gt;&lt;/b&gt;&lt;/div&gt;&lt;div&gt;&lt;br&gt;&lt;/div&gt;&lt;div&gt;{}, {}&lt;/div&gt;&lt;div&gt;Since: {}&lt;/div&gt;" style="rounded=1;whiteSpace=wrap;html=1;align=left;spacingLeft=0;spacingTop=0;spacing=10;fontSize=16;" vertex="1" parent="{}">
              <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
            </mxCell>"#,
                    calling_num,
                    sanitize(&calling.name),
                    member.last_name,
                    member.given_names,
                    member.held_calling_since.to_string()?,
                    sub_org_bubble_id,
                    calling_bubble_cursor_x,
                    calling_bubble_cursor_y,
                    conf.calling_bubble_dimensions.width,
                    conf.calling_bubble_dimensions.height,
                );
                calling_num += 1;

                sub_org_bubble_elems.push(calling_bubble_elem);

                i += 1;
            }

            // The first element in sub_org_calling_bubble_elems is a placeholder for the sub-org bubble itself,
            // so subtract one from the length of the sub_org_bubble_elems list to do this calculation
            let sub_org_filled_callings_count = sub_org_bubble_elems.len() - 1;
            let mut sub_org_bubble_height = ((sub_org_filled_callings_count - 1)
                / conf.calling_bubbles_per_row as usize
                + 1) as i32
                * (conf.calling_bubble_dimensions.height + conf.calling_bubble_vertical_margins)
                + conf.calling_bubble_vertical_margins
                + conf.org_bubble_title_height;
            if sub_org_bubble_height < min_org_bubble_height {
                sub_org_bubble_height = min_org_bubble_height;
            }

            sub_org_bubble_cursor_y += conf.sub_org_bubble_vertical_margins;

            let sub_org_bubble_elem = format!(
                r#"          <mxCell id="{}" value="&lt;font style=&quot;font-size: 22px;&quot;&gt;{}&lt;/font&gt;" style="swimlane;whiteSpace=wrap;html=1;rounded=1;strokeWidth=4;startSize=40;strokeColor=#9E9E9E;fontColor=#6B6B6B;" vertex="1" parent="{}">
            <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
          </mxCell>"#,
                sub_org_bubble_id,
                child_name,
                org_bubble_id,
                conf.sub_org_bubble_horzontal_margins,
                sub_org_bubble_cursor_y,
                sub_org_bubble_width,
                sub_org_bubble_height,
            );

            sub_org_bubble_elems[0] = sub_org_bubble_elem;

            sub_org_bubble_cursor_y += sub_org_bubble_height;

            all_sub_orgs_in_org_bubble_elems.push(sub_org_bubble_elems);
        }

        let org_filled_callings_count = org_calling_bubble_elems.len();
        let mut org_bubble_height = if org.children.is_empty() {
            ((org_filled_callings_count - 1) / conf.calling_bubbles_per_row as usize + 1) as i32
                * (conf.calling_bubble_dimensions.height + conf.calling_bubble_vertical_margins)
                + conf.calling_bubble_vertical_margins
                + conf.org_bubble_title_height
        } else {
            sub_org_bubble_cursor_y + conf.sub_org_bubble_vertical_margins
        };

        if org_bubble_height < min_org_bubble_height {
            org_bubble_height = min_org_bubble_height;
        }

        org_bubble_cursor_y += org_bubble_height;

        let org_bubble_elem = format!(
            r#"        <mxCell id="{}" value="&lt;font style=&quot;font-size: 22px;&quot;&gt;{}&lt;/font&gt;" style="swimlane;whiteSpace=wrap;html=1;rounded=1;strokeWidth=4;startSize=40;" vertex="1" parent="1">
          <mxGeometry x="{}" y="{}" width="{}" height="{}" as="geometry" />
        </mxCell>"#,
            org_bubble_id,
            org_name,
            org_bubble_x,
            org_bubble_y,
            conf.org_bubble_width,
            org_bubble_height,
        );

        file_contents.push_str(&org_bubble_elem);

        for elem in org_calling_bubble_elems {
            file_contents.push_str(&elem);
        }

        for sub_org_bubble_elems in all_sub_orgs_in_org_bubble_elems {
            for elem in sub_org_bubble_elems {
                file_contents.push_str(&elem);
            }
        }

        org_bubble_cursor_y += conf.org_bubble_margins.bottom;
    }

    let diagram_footer = r#"      </root>
    </mxGraphModel>
  </diagram>
</mxfile>"#;

    file_contents.push_str(diagram_footer);

    Ok(file_contents)
}

fn sanitize(str: &str) -> String {
    str.replace("&", "and")
}

fn rand_tag(rng: &mut impl Rng) -> u32 {
    rng.gen_range(100000..1000000)
}
